### ISSUE TITLE
HTML tags in bootstrapped JSON cause JavaScript parsing error, breaking builder

### ISSUE THREAD

**Issue by ghing** [denverfoundation/storybase#658]
When a script tag is in an asset's body, and it is rendered in the bootstrapped data for the story builder by `StoryBuilderVIew`/`story_builder.html`, the browser interprets the closing `</script>` in the asset body as the closing of the containing script tag.

This problem is described in this Stack Overflow question: [Script tag in JavaScript string](http://stackoverflow.com/questions/1659749/script-tag-in-javascript-string)

Error messages look like this in Chrome:

```
[Django] ERROR: javascript error: {"HTTP_USER_AGENT": "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.22 (KHTML, like Gecko) Chrome/25.0.1364.97 Safari/537.22", "url": "http://floodlightproject.org/en/build/STORY_ID_HERE/", "user": "XXXXXX@gmail.com", "lineNumber": "681", "message": "Uncaught SyntaxError: Unexpected token ILLEGAL", "location": "http://floodlightproject.org/en/build/STORY_ID_HERE/"}
```

and this in Firefox:

```
[Django] ERROR: javascript error: {"HTTP_USER_AGENT": "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:18.0) Gecko/20100101 Firefox/18.0", "url": "http://floodlightproject.org/en/build/STORY_ID_HERE/", "user": "XXXXXX@gmail.com", "lineNumber": "681", "message": "SyntaxError: unterminated string literal", "location": "http://floodlightproject.org/en/build/STORY_ID_HERE/"}
```

An example of the offending JSON string is:

```
        {
            "asset_created": null,
            "asset_id": "ASSET_ID_REDACTED",
            "attribution": "Source: Newspaper Association of America\t\t\n\t\t\nPew Research Center's Project for Excellence in Journalism\t\t\n2012 State of the News Media\t\t\n",
            "body": "<script src=\"https://docs.google.com/spreadsheet/gpub?url=http%3A%2F%2Foj0ijfii34kccq3ioto7mdspc7r2s7o9-ss-opensocial.googleusercontent.com%2Fgadgets%2Fifr%3Fup_title%3D11-Year%2520Decline%2520for%2520Daily%2520Advertising%2520Revenue%26up_initialstate%3D%257B%2522time%2522%253A%25222011%2522%252C%2522yLambda%2522%253A1%252C%2522iconKeySettings%2522%253A%255B%255D%252C%2522xZoomedIn%2522%253Afalse%252C%2522xZoomedDataMin%2522%253A473385600000%252C%2522yZoomedIn%2522%253Afalse%252C%2522yAxisOption%2522%253A%25225%2522%252C%2522playDuration%2522%253A15000%252C%2522sizeOption%2522%253A%2522_UNISIZE%2522%252C%2522orderedByX%2522%253Afalse%252C%2522xAxisOption%2522%253A%2522_TIME%2522%252C%2522colorOption%2522%253A%25224%2522%252C%2522duration%2522%253A%257B%2522multiplier%2522%253A1%252C%2522timeUnit%2522%253A%2522Y%2522%257D%252C%2522xZoomedDataMax%2522%253A1293840000000%252C%2522dimensions%2522%253A%257B%2522iconDimensions%2522%253A%255B%2522dim0%2522%255D%257D%252C%2522iconType%2522%253A%2522LINE%2522%252C%2522nonSelectedAlpha%2522%253A0.4%252C%2522xLambda%2522%253A1%252C%2522uniColorForNonSelected%2522%253Afalse%252C%2522yZoomedDataMin%2522%253A0%252C%2522yZoomedDataMax%2522%253A50%252C%2522orderedByY%2522%253Afalse%252C%2522showTrails%2522%253Afalse%257D%26up__table_query_url%3Dhttps%253A%252F%252Fdocs.google.com%252Fspreadsheet%252Ftq%253Frange%253DA5%25253AF32%2526gid%253D0%2526key%253D0AhWnhIC_xoBpdHlqMjBzSTVsNldBb29NY1JGTWdNZHc%2526pub%253D1%26url%3Dhttp%253A%252F%252Fwww.google.com%252Fig%252Fmodules%252Fmotionchart.xml%26spreadsheets%3Dspreadsheets&height=667&width=1821\"></script>\nDone",
            "caption": "",
            "content": "<figure>\n<script src=\"https://docs.google.com/spreadsheet/gpub?url=http%3A%2F%2Foj0ijfii34kccq3ioto7mdspc7r2s7o9-ss-opensocial.googleusercontent.com%2Fgadgets%2Fifr%3Fup_title%3D11-Year%2520Decline%2520for%2520Daily%2520Advertising%2520Revenue%26up_initialstate%3D%257B%2522time%2522%253A%25222011%2522%252C%2522yLambda%2522%253A1%252C%2522iconKeySettings%2522%253A%255B%255D%252C%2522xZoomedIn%2522%253Afalse%252C%2522xZoomedDataMin%2522%253A473385600000%252C%2522yZoomedIn%2522%253Afalse%252C%2522yAxisOption%2522%253A%25225%2522%252C%2522playDuration%2522%253A15000%252C%2522sizeOption%2522%253A%2522_UNISIZE%2522%252C%2522orderedByX%2522%253Afalse%252C%2522xAxisOption%2522%253A%2522_TIME%2522%252C%2522colorOption%2522%253A%25224%2522%252C%2522duration%2522%253A%257B%2522multiplier%2522%253A1%252C%2522timeUnit%2522%253A%2522Y%2522%257D%252C%2522xZoomedDataMax%2522%253A1293840000000%252C%2522dimensions%2522%253A%257B%2522iconDimensions%2522%253A%255B%2522dim0%2522%255D%257D%252C%2522iconType%2522%253A%2522LINE%2522%252C%2522nonSelectedAlpha%2522%253A0.4%252C%2522xLambda%2522%253A1%252C%2522uniColorForNonSelected%2522%253Afalse%252C%2522yZoomedDataMin%2522%253A0%252C%2522yZoomedDataMax%2522%253A50%252C%2522orderedByY%2522%253Afalse%252C%2522showTrails%2522%253Afalse%257D%26up__table_query_url%3Dhttps%253A%252F%252Fdocs.google.com%252Fspreadsheet%252Ftq%253Frange%253DA5%25253AF32%2526gid%253D0%2526key%253D0AhWnhIC_xoBpdHlqMjBzSTVsNldBb29NY1JGTWdNZHc%2526pub%253D1%26url%3Dhttp%253A%252F%252Fwww.google.com%252Fig%252Fmodules%252Fmotionchart.xml%26spreadsheets%3Dspreadsheets&height=667&width=1821\"></script>\nDone\n<figcaption><div class='attribution'>Source: Newspaper Association of America\t\t\n\t\t\nPew Research Center's Project for Excellence in Journalism\t\t\n2012 State of the News Media\t\t\n</div></figcaption>\n</figure>",
            "created": "2013-02-01T15:34:19.357351",
            "display_title": "Done",
            "image": null,
            "language": "en",
            "languages": [
                {
                    "id": "en",
                    "name": "English",
                    "url": "/en/assets/0134435e58034940af508afccd692224/"
                }
            ],
            "last_edited": "2013-02-01T15:34:19.891458",
            "license": "CC BY",
            "published": null,
            "resource_uri": "/api/0.1/assets/0134435e58034940af508afccd692224/",
            "section_specific": false,
            "source_url": "",
            "status": "draft",
            "thumbnail_url": null,
            "title": "",
            "type": "chart",
            "url": null
        }
```

It's easier to understand what's going on with this dummy example, with a shorter script `src` attribute:

```
        {
            "asset_created": null,
            "asset_id": "0134435e58034940af508afccd692224",
            "attribution": "Source: Newspaper Association of America\t\t\n\t\t\nPew Research Center's Project for Excellence in Journalism\t\t\n2012 State of the News Media\t\t\n",
            "body": "<script src=\"http://fake.floodlightproject.org/script.js\"></script>\nDone",
            "caption": "",
            "content": "<figure>\n<script src=\"http://fake.floodlightproject.org/script.js\"></script>\nDone\n<figcaption><div class='attribution'>Source: Newspaper Association of America\t\t\n\t\t\nPew Research Center's Project for Excellence in Journalism\t\t\n2012 State of the News Media\t\t\n</div></figcaption>\n</figure>",
            "created": "2013-02-01T15:34:19.357351",
            "display_title": "Done",
            "image": null,
            "language": "en",
            "languages": [
                {
                    "id": "en",
                    "name": "English",
                    "url": "/en/assets/0134435e58034940af508afccd692224/"
                }
            ],
            "last_edited": "2013-02-01T15:34:19.891458",
            "license": "CC BY",
            "published": null,
            "resource_uri": "/api/0.1/assets/0134435e58034940af508afccd692224/",
            "section_specific": false,
            "source_url": "",
            "status": "draft",
            "thumbnail_url": null,
            "title": "",
            "type": "chart",
            "url": null
        }
```

In particular, it's the closing `</script>` tag in the `body` and `content` properties of the JSON object:

```
"body": "<script src=\"http://fake.floodlightproject.org/script.js\"></script>\nDone",
```

Note that all the quotes are escaped properly and that the JSON object validates.

In general, allowing users to enter `<script>` tags and then rendering them isn't great from a security standpoint anyway.  I think the ultimate solution would be to prohibit `<script>` tags in assets and direct users to use oEmbed when possible, or fake oEmbed with our own endpoints for services that don't support it.

**Comment by ghing** [comment]
Updated the title of the issue to more accurately reflect the problem.

**Comment by ghing** [comment]
This HTML, as reported by a user, in an asset also causes the issue.  Note that it doesn't contain any `<script>` tags. I'm trying to figure out what about this markup is causing the problems.

```
<p>Here are the
facts: </p>
<p>1.&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <strong>1 in 7 Students in Grades K-12 is
either a bully or a victim of bullying.</strong><strong> </strong><span><br>
<br>
<strong>2.&nbsp;&nbsp;&nbsp;&nbsp; </strong>90%<strong> of 4th through
8th graders report being victims of bullying</strong><strong></strong></span></p>

<strong>3.&nbsp;&nbsp;&nbsp;&nbsp; </strong>1 out of every 10<strong> students who drops out of school because of </strong>repeated bullying.
```

Trying to diagnose what about the HTML breaks things.  This also breaks things:

```
<p>Here are the
facts: </p>
<p>1.&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <strong>1 in 7 Students in Grades K-12 is
either a bully or a victim of bullying.</strong><strong> 
</strong></p>
```

but this doesn't:

```
<p>Here are the
facts: </p>
<p>1.&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <strong>1 in 7 Students in Grades K-12 is
either a bully or a victim of bullying.</strong><strong> </strong></p>
```

but weirdly, this doesn't break either:

```
<p><strong>
</strong></p>
```

I'm not going to spend any more time on this for now...

**Comment by ghing** [comment]
There's a lengthy discussion of dealing with this very situation in Rails at https://github.com/rails/rails/pull/3354.  If I'm parsing the discussion carefully, it looks like the JSON spec parses `/` and `\/` the same way to avoid this very problem.  We just need to escape our "solidus" character in our HTML.  I think this is as simple as

```
json_string = '[{"asset_created": null, "asset_id": "575155948f7d4c4ebccb02d4e8f84d2f", "body": "<script></script>"}]'
escaped_json_string = json_string.replace('</', r'<\/')
```

but I [asked on Stack Overflow](http://stackoverflow.com/questions/15297028/how-do-i-escape-closing-in-html-tags-in-json-with-python) to see if there's something I'm overlooking.

If we just do this escaping before returning the JSON at the end of `StoryBuilderView.get_assets_json` and `StoryBuilderView.get_section_assets_json` we should be good.

**Comment by ghing** [comment]
Moving this to Iteration 12 because it seems pretty high priority.

**Comment by ghing** [comment]
[simplejson.JSONEncoderforHTML](http://simplejson.readthedocs.org/en/latest/#simplejson.JSONEncoderForHTML) does this escaping. By default, Tastypie uses `core.serializers.json.DjangoJSONEncoder` to serialize JSON data.  It doesn't do any escaping.  The version of simplejson that ships with Django doesn't provide JSONEncoderForHTML.

This is the escaping code that JSONEncoderForHTML uses:

```

    def iterencode(self, o, _one_shot=False):
        chunks = super(JSONEncoderForHTML, self).iterencode(o, _one_shot)
        for chunk in chunks:
            chunk = chunk.replace('&', '\\u0026')
            chunk = chunk.replace('<', '\\u003c')
            chunk = chunk.replace('>', '\\u003e')
            yield chunk
```

**Comment by ghing** [comment]
Ok, the HTML fragment that's breaking things as described in https://github.com/PitonFoundation/atlas/issues/658#issuecomment-14593189 is not due to "</" being contained in the JSON.  It has to do with some weird unicode characters.  I'll open a new issue for it.

**Comment by ghing** [comment]
Closed as of 9e292270270a00345080db5d5469f0e0089c6e2b
